### DISCUSSION TITLE
How can I re encode photos to HEIC on Windows?

### DISCUSSION THREAD

**Discussion by YousufSSyed** [ImageMagick/ImageMagick#1966] [General]
I'm a beginner with ImageMagick. I've tried commands like `magick filename.jpg filename.heic`, but when I do `magick inspect filename.heic`, it returns `filename.heic JPEG`. When I entered `identify -list format`, it shows heic* (*native blob support). I heard you need libheif but I'm sure what to do. How can I setup Imagemagick on Windows so that I can re encode photos to HEIC?

**Comment by fmw42** [comment]
I am not a Windows user so cannot advise about installation. But you need the libheif delegate installed. See https://github.com/strukturag/libheif

**Comment by yaleman** [comment]
# edit: I'm a derp, this only works if you have the proper libraries installed as per @fmw42's comment above. I'll leave my original comment below:

The correct command is `magick convert filename.jpg filename.heic`.

[The convert command documentation](https://imagemagick.org/script/convert.php)

This is tested working with the [base windows binary install](https://imagemagick.org/script/download.php#windows)

**Reply by YousufSSyed** [reply]
I downloaded the latest Imagemagick (the top link in the windows section) and this command still didn't work. When I used `magick identify` on the output images, I still got JPEG.

**Reply by yaleman** [reply]
Yeah, I just realised the previous commenter was right, my deepest apologies. I'd been rather mistaken in my own usage for some time.

`magick -list format | findstr /I heic` shows that it can only read the format.

**Reply by fmw42** [reply]
By the way, on IM 7, use magick only not magick convert otherwise, you will get IM 6 behavior

**Comment by yaleman** [comment]
This was bugging me so I worked it out... after finding a reference on [this thread in the libheif library issues](https://github.com/strukturag/libheif/issues/200) I downloaded the libheif windows binaries built by @pphh77.

* releases page: https://github.com/pphh77/libheif-Windowsbinary/releases
* direct link: https://github.com/pphh77/libheif-Windowsbinary/releases/download/v1.6.2/libheif-1.6.2-r1106-g3824054.7z

Extracted the archive to (in my example): `E:\Zoo\libheif-1.6.2-r1106-g3824054\x64\`

The binary will thus be at: `E:\Zoo\libheif-1.6.2-r1106-g3824054\x64\heif-enc.exe`

I'm using the HDR x64 install of ImageMagic so I edited the xml file: `C:\Program Files\ImageMagick-7.0.10-Q16-HDRI\delegates.xml`. Find `delegates.xml` in your install directory either way.

I added the following lines to support PNG and JPEG conversion:

```
  <delegate decode="jpg" encode="heic" command="cmd.exe /c (move &quot;%i&quot; &quot;%i.jpg&quot; >nul) &amp; (E:/Zoo/libheif-1.6.2-r1106-g3824054/x64/heif-enc.exe &quot;%i.jpg&quot;) &amp; (move &quot;%i.heic&quot; &quot;%o&quot; >nul) &amp; (move &quot;%i.jpg&quot; &quot;%i&quot; >nul)" />
  <delegate decode="png" encode="heic" command="cmd.exe /c (move &quot;%i&quot; &quot;%i.jpg&quot; >nul) &amp; (E:/Zoo/libheif-1.6.2-r1106-g3824054/x64/heif-enc.exe &quot;%i.png&quot;) &amp; (move &quot;%i.heic&quot; &quot;%o&quot; >nul) &amp; (move &quot;%i.png&quot; &quot;%i&quot; >nul)" />
```

It's a bit hacky but you can add `-q n` after `heif-enc.exe` to specify a HEIF quality. (ie, `-q 75` sets HEIF quality to 75.

**Reply by YousufSSyed** [reply]
I tried this on my PC and it works. Thanks!

**Reply by walterboyd2** [reply]
Yaleman, well done!  I'm using your app directly, because something funny going on if I run it through imageMagick.  The resulting file is 4x larger than if I use heif-enc.exe to perform the conversion.

`magick x.png -quality 30 xTest1.heic`  ===> resulting file is 108k

`heif-enc -q 30 -o xTest2.heic x.png`  ===> resulting file is 24k.  And it looks better!  xTest1 has a lot of jpeg looking artifacts.

**Comment by snibgo** [comment]
As yaleman mentioned, the suggested lines in delegates.xml do not have a "-q N" setting. In the command "magick x.png -quality 30 xTest1.heic", the "-quality" setting has no effect.

(Edit: Or maybe IM "-quality" does have an effect. See comments below. I have no experience with HEIC.)

**Reply by walterboyd2** [reply]
Thanks, I missed that.  I see now that yaleman intended "-q N" would actually be hard-coded to some value in delegates.  Makes sense now.

But -quality is having an effect; I think it's causing imageMagick to use jpg compression.

**Reply by yaleman** [reply]
I think that might be setting the quality in imagemagick as the delegate basically exports to jpeg and makes the conversion
